File: deep_search-dev/test_deepsearch_api.py
import asyncio
import aiohttp
import os
from logs.logger import define_log_level


async def fetch(session, topic, index, total, logger):
    # url = 'http://10.176.238.83:7396/api/deep_search'
    url = 'http://localhost:7396/api/deep_search'
    logger.info(f'ip:{url}')
    logger.info(f'【处理中】第 {index + 1}/{total} 条数据: {topic}')
    for i in range(3):  # 3 次机会
        try:
            async with session.post(url, json={"topic": topic, 'lang': 'en'}) as response:
                if response.status == 200:
                    result = await response.text()
                    if eval(result)['status'] == 200:
                        logger.info(f'✅ 完成 第 {index + 1} 条: {topic} | 响应长度: {len(result)}，{result[:50]}...')
                        break
                    else:
                        logger.error(f'❌ 完成 第 {index + 1} 条: {topic} | 响应长度: {len(result)}，{result[:50]}...')
                else:
                    logger.error(f'❌ 第 {index + 1} 条: {topic} 请求失败，{response.text()[:50]}...')
        except Exception as e:
            logger.error(f'⚠️ 第 {index + 1} 条: {topic} 请求失败，出错: {str(e)}')
            pass


async def worker(session, topic, index, total, semaphore, logger):
    async with semaphore:
        await fetch(session, topic, index, total, logger)


async def main(topics, max_concurrent=5):
    # 初始化日志
    project_root = os.path.dirname(os.path.abspath(__file__))
    logger, log_name = define_log_level(project_root, "test_deepsearch_api")
    logger.info("开始测试deepsearch API...")
    
    total = len(topics)
    semaphore = asyncio.Semaphore(max_concurrent)  # 控制最大并发数量
    async with aiohttp.ClientSession() as session:
        tasks = [
            worker(session, topic, i, total, semaphore, logger)
            for i, topic in enumerate(topics)
        ]
        await asyncio.gather(*tasks)


if __name__ == '__main__':
    topics = ['最近医疗大健康有什么新动态，对创业者，个人有什么机遇？']
    asyncio.run(main(topics))

File: deep_search-dev-go/test_deepsearch_api.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"sync"
	"time"

	"deepsearch/logs"
)

// 请求结构体
type Request struct {
	Topic string `json:"topic"`
	Lang  string `json:"lang"`
}

// 响应结构体
type Response struct {
	Status int `json:"status"`
}

// fetch 函数模拟Python版本的fetch函数
func fetch(client *http.Client, topic string, index int, total int, logger *logs.Logger, wg *sync.WaitGroup, semaphore chan struct{}) {
	defer wg.Done()

	// 获取信号量
	semaphore <- struct{}{}
	defer func() { <-semaphore }()

	// API URL
	url := "http://localhost:7396/api/deep_search"
	logger.Info("ip:" + url)
	logger.Info(fmt.Sprintf("【处理中】第 %d/%d 条数据: %s", index+1, total, topic))

	// 3次重试机会
	for i := 0; i < 3; i++ {
		// 创建请求体
		requestBody := Request{
			Topic: topic,
			Lang:  "en",
		}
		jsonData, err := json.Marshal(requestBody)
		if err != nil {
			logger.Error(fmt.Sprintf("⚠️ 第 %d 条: %s 请求失败，出错: %v", index+1, topic, err))
			continue
		}

		// 发送POST请求
		req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
		if err != nil {
			logger.Error(fmt.Sprintf("⚠️ 第 %d 条: %s 请求失败，出错: %v", index+1, topic, err))
			continue
		}
		req.Header.Set("Content-Type", "application/json")

		resp, err := client.Do(req)
		if err != nil {
			logger.Error(fmt.Sprintf("⚠️ 第 %d 条: %s 请求失败，出错: %v", index+1, topic, err))
			continue
		}

		defer resp.Body.Close()

		// 读取响应
		body, err := io.ReadAll(resp.Body)
		if err != nil {
			logger.Error(fmt.Sprintf("⚠️ 第 %d 条: %s 请求失败，出错: %v", index+1, topic, err))
			continue
		}

		result := string(body)

		if resp.StatusCode == 200 {
			// 解析JSON响应
			var response Response
			err := json.Unmarshal(body, &response)
			if err != nil {
				logger.Error(fmt.Sprintf("⚠️ 第 %d 条: %s 解析响应失败，出错: %v", index+1, topic, err))
				continue
			}

			if response.Status == 200 {
				// 截取前50个字符用于日志
				shortResult := result
				if len(shortResult) > 50 {
					shortResult = shortResult[:50]
				}
				logger.Info(fmt.Sprintf("✅ 完成 第 %d 条: %s | 响应长度: %d，%s...", index+1, topic, len(result), shortResult))
				return
			} else {
				// 截取前50个字符用于日志
				shortResult := result
				if len(shortResult) > 50 {
					shortResult = shortResult[:50]
				}
				logger.Error(fmt.Sprintf("❌ 完成 第 %d 条: %s | 响应长度: %d，%s...", index+1, topic, len(result), shortResult))
				return
			}
		} else {
			// 截取前50个字符用于日志
			shortResult := result
			// if len(shortResult) > 50 {
			// 	shortResult = shortResult[:50]
			// }
			logger.Error(fmt.Sprintf("❌ 第 %d 条: %s 请求失败，%s...", index+1, topic, shortResult))
			continue
		}
	}
}

func main() {
	// 初始化日志
	projectRoot := filepath.Dir(os.Args[0])
	logger, _ := logs.DefineLogLevel(projectRoot, "test_deepsearch_api", "INFO", "INFO")
	logger.Info("开始测试deepsearch API...")

	// 测试主题
	topics := []string{"最近医疗大健康有什么新动态，对创业者，个人有什么机遇？"}
	total := len(topics)

	// 控制最大并发数量
	maxConcurrent := 5
	semaphore := make(chan struct{}, maxConcurrent)

	// 创建HTTP客户端
	client := &http.Client{
		Timeout: 300 * time.Second, // 增加到5分钟
	}

	// 使用WaitGroup等待所有goroutine完成
	var wg sync.WaitGroup

	// 启动goroutine处理每个主题
	for i, topic := range topics {
		wg.Add(1)
		go fetch(client, topic, i, total, logger, &wg, semaphore)
	}

	// 等待所有goroutine完成
	wg.Wait()
	logger.Info("所有测试已完成")
}
